pageserver/src/tenant/storage_layer/layer.rs

```diff
@@ -0,0 +1,1139 @@
+        file_name: LayerFileName,
+        metadata: LayerFileMetadata,
+    ) -> ResidentLayer {
+        let desc = PersistentLayerDesc::from_filename(
+            timeline.tenant_id,
+            timeline.timeline_id,
+            file_name,
+            metadata.file_size(),
+        );
+
+        let access_stats = LayerAccessStats::for_loading_layer(LayerResidenceStatus::Resident);
+
+        let mut resident = None;
+
+        let owner = Layer(Arc::new_cyclic(|owner| {
+            let inner = Arc::new(DownloadedLayer {
+                owner: owner.clone(),
+                kind: tokio::sync::OnceCell::default(),
+            });
+            resident = Some(inner.clone());
+
+            LayerInner::new(conf, timeline, access_stats, desc, Some(inner))
+        }));
+
+        debug_assert!(owner.0.needs_download_blocking().unwrap().is_none());
+
+        let downloaded = resident.expect("just initialized");
+
+        ResidentLayer { downloaded, owner }
+    }
+
+    /// Creates a Layer value for freshly written out new layer file by renaming it from a
+    /// temporary path.
+    pub(crate) fn for_written_tempfile(
+        conf: &'static PageServerConf,
+        timeline: &Arc<Timeline>,
+        desc: PersistentLayerDesc,
+        temp_path: &Path,
+    ) -> anyhow::Result<ResidentLayer> {
+        let mut resident = None;
+
+        let owner = Layer(Arc::new_cyclic(|owner| {
+            let inner = Arc::new(DownloadedLayer {
+                owner: owner.clone(),
+                kind: tokio::sync::OnceCell::default(),
+            });
+            resident = Some(inner.clone());
+            let access_stats = LayerAccessStats::empty_will_record_residence_event_later();
+            access_stats.record_residence_event(
+                LayerResidenceStatus::Resident,
+                LayerResidenceEventReason::LayerCreate,
+            );
+            LayerInner::new(conf, timeline, access_stats, desc, Some(inner))
+        }));
+
+        let downloaded = resident.expect("just initialized");
+
+        // if the rename works, the path is as expected
+        std::fs::rename(temp_path, owner.local_path())
+            .context("rename temporary file as correct path for {owner}")?;
+
+        Ok(ResidentLayer { downloaded, owner })
+    }
+
+    /// Requests the layer to be evicted and waits for this to be done.
+    ///
+    /// If the file is not resident, an [`EvictionError::NotFound`] is returned.
+    ///
+    /// If for a bad luck or blocking of the executor, we miss the actual eviction and the layer is
+    /// re-downloaded, [`EvictionError::Downloaded`] is returned.
+    pub(crate) async fn evict_and_wait(
+        &self,
+        rtc: &RemoteTimelineClient,
+    ) -> Result<(), EvictionError> {
+        self.0.evict_and_wait(rtc).await
+    }
+
+    /// Delete the layer file when the `self` gets dropped, also schedule a remote index upload
+    /// then perhaps.
+    pub(crate) fn garbage_collect(&self) {
```

Comment:
Please pick a different function name. Garbage collection already has domain specific meaning inside pageserver (`gc_timeline`).

Also I commented about the name at one of the call sites.

Reply:
https://github.com/neondatabase/neon/pull/4938#discussion_r1307312511